### DISCUSSION TITLE
Loading an entity twice on failure

### DISCUSSION THREAD

**Discussion by danielwinkler** [ravendb/ravendb#16227] [Q&A] [answered]
Dear all,

I would like to try to load an entity again in case that the request fails.
As much as i can see from the code, it seems that `_knownMissingIds` is causing this test to fail.

```csharp

    [Fact]
    public void ShouldLoadTwiceIfTwoTimesNull()
    {
        var id = "idontexist";
        using (var store = GetDocumentStore())
        {
            using var session = store.OpenSession();
            var l1 = session.Load<Level>(id);
            Assert.Null(l1);
            Assert.Equal(1, session.Advanced.NumberOfRequests);
            var (l2, cv) = session.Advanced.ConditionalLoad<Level>(id, string.Empty);
            Assert.Null(l2);
            Assert.Null(cv);
            Assert.Equal(2, session.Advanced.NumberOfRequests);
        }
    }
```

Is there a way i can force to load again, or this is only an issue for the in memory session?

see [RavenDB-15041](https://github.com/ravendb/ravendb/pull/11085)

**Comment by georgiosd** [comment]
`session.Advanced.Refresh` would probably do it. But why do you want to do this?

**Reply by danielwinkler** [reply]
Thanks @georgiosd , I thought of that, but if the entity is null then i have nothing to pass into refresh, except `null`

My actual problem is different:

This is a workaround for a slow replication issue.
So
- User creates doc-1 on node A
- User client receives doc-1 DTO and immediately sends another request to the backend
- backend reads by chance from node B
- replication did not happen
- backend receives a null entity
- backend retries after some timeout to load doc-1 again

I think i will use `WaitForReplicationAfterSaveChanges` to avoid the problem in the first place, or use a clusterwide transaction.

But still i would be curious to know if i can force a reload.

**Reply by georgiosd** [reply]
Ah yes! Funny, I was just dealing with a similar race condition, but with an attachment.

You could also use the document store directly and trigger the load - if it works the way you want you can just store the object to the session with its change vector to track it.

I was doing this in the context of subscriptions, so I just used `@refresh` to try again later.

**Reply by ayende** [reply]
The better way to do that is to abort handling of the entire message, and retry afterward.
Another would be to use a separate session.

**Reply by danielwinkler** [reply]
thanks @ayende, @georgiosd! so the cleanest solution would be to open a fallback session, try to load in the new session, and if successful store it in the main session. 

To get around the issue i would configure `WaitForReplicationAfterSaveChanges` to not `throwOnTimeout`. Is there a way to only replicate to the currently healthy nodes, otherwise I'd use `session.Advanced.DocumentStore.Urls.Length - 1` as value for replicas.

```csharp
            var clusterCount = session.Advanced.DocumentStore.Urls.Length;
            session.Advanced.WaitForReplicationAfterSaveChanges(
                timeout: TimeSpan.FromSeconds(3),
                throwOnTimeout: false,
                replicas: clusterCount - 1,
                majority: true
            );
```

But, if one node is unhealthy all SaveChanges will now take 3 seconds if I am not mistaken.
On the other hand, the unhealthy node will not be used to load so i would not have to wait on the save operation.

**Reply by ayende** [reply]
Well' start from the basic, why are you hitting Node B on a regular basis anyway?
All clients should be hitting the same node all the time by default.

Are you using RoundRobin?

**Reply by danielwinkler** [reply]
That we are hitting node B is my only explanation for the behavior, but maybe I'm wrong.
It's not really regular, but happens sometimes when we are running test.
We are not using RoundRobin and not FastestNode, we are using default.

**Reply by ayende** [reply]
The client is waiting for the operation to complete before?
Can you show the traffic watch when you reproduce this?

**Reply by danielwinkler** [reply]
Hi @ayende ,

Yes the client is waiting, see the traffic watch below:

We are getting this on node A
```json

  {
    "TrafficWatchType": "Http",
    "TimeStamp": "2023-04-11T21:27:56.5597186Z",
    "DatabaseName": "cluster-dev",
    "CustomInfo": "sessions-42873-B",
    "ClientIP": "35.204.119.79",
    "CertificateThumbprint": "60C136561EB4C0476CE8F731EB1D25299C03B16D",
    "RequestId": 38953,
    "HttpMethod": "GET",
    "ElapsedMilliseconds": 0,
    "ResponseStatusCode": 404,
    "RequestUri": "https://a.cluster-dev.corp.ravendb.cloud/databases/cluster-dev/docs?&include=AssignmentId&include=TrainingId&id=sessions-42873-B",
    "AbsoluteUri": "https://a.cluster-dev.corp.ravendb.cloud",
    "RequestSizeInBytes": 0,
    "ResponseSizeInBytes": 0,
    "Type": "Documents"
  }
```

and this on node B
```json
  {
    "TrafficWatchType": "Http",
    "TimeStamp": "2023-04-11T21:27:56.4132644Z",
    "DatabaseName": "cluster-dev",
    "CustomInfo": "PUT    sessions-42873-B\n",
    "ClientIP": "35.204.119.79",
    "CertificateThumbprint": "60C136561EB4C0476CE8F731EB1D25299C03B16D",
    "RequestId": 4064,
    "HttpMethod": "POST",
    "ElapsedMilliseconds": 44,
    "ResponseStatusCode": 201,
    "RequestUri": "https://b.cluster-dev.corp.ravendb.cloud/databases/cluster-dev/bulk_docs?",
    "AbsoluteUri": "https://b.cluster-dev.corp.ravendb.cloud",
    "RequestSizeInBytes": 449,
    "ResponseSizeInBytes": 348,
    "Type": "BulkDocs"
  },
```

Our cluster is [Node A, Node B, Node C], where Node B is the leader.

**Reply by danielwinkler** [reply]
Hi @ayende,

is there any other information that I could provide?

Cheers, Daniel

**Reply by ayende** [reply, marked as answer]
Hi,
This looks really strange, because if your configuration is not round robin, all operations should go to node B.
We fixed this issue: https://github.com/ravendb/ravendb/pull/16121

Which may expose similar behavior.

If you restart the _client_ nodes on both, do you still see the same behavior?

**Reply by danielwinkler** [reply]
Thanks for your reply.

It indeed got much better after the restart, but after some time i was able to reproduce it again. Can you recommend me something for debugging the cluster configuration on the node?

One thing would be listening to Topology updated events, the other maybe to inspect the urls before every Load/Store operation?

**Reply by ayende** [reply]
What client version are you using?

**Reply by danielwinkler** [reply]
Hi @ayende 

```
<PackageReference Include="RavenDB.Client" Version="5.4.103" />
```

My theory would be as follows:
- When we are running our scripts we are putting quite some load on the indexes, so the cpu utilization maxes out.
- Because of the high load, one of the DB clients cannot reach the first node of the topology
- The client will fail over to the second node, and perform the request at node 2
- The client will stay on this node until the spawned health check responds successfully from node 1 (?)

If these assumptions are somewhat correct:

- Is there a way to see which node is currently the preferred node, and which nodes had errors?
- what options besides write assurance do have to avoid the original issue?

**Reply by ayende** [reply]
The fix in question is after that release. Can you move to a nightly build for the client in question?

The situation there is that a failure like you describe wouldn't be reverted back to the owner node, leading to a split brain situation

**Reply by ayende** [reply]
Alternatively, this will be in the next stable build, expected mid may

**Reply by danielwinkler** [reply]
Thanks @ayende , i think we were running into the split brain situation.

Where are the client nightlies published?, I only found the server,

**Reply by garayx** [reply]
the client nightlies are published here: https://myget.org/feed/ravendb/package/nuget/RavenDB.Client
there is `Package history` section on bottom of the page
